### ISSUE TITLE
Model.predict gives error when using preprocessing

### ISSUE THREAD

**Issue by Staano** [tflearn/tflearn#815]
I'm applying the following preprocessing on the training images

```
img_prep = ImagePreprocessing()
img_prep.add_featurewise_zero_center() samples.
img_prep.add_featurewise_stdnorm() 

```

When trying use `model.predict()` it gives the following error:

```
/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/site-packages/tflearn/data_preprocessing.pyc in _featurewise_zero_center(self, batch)
    214     def _featurewise_zero_center(self, batch):
    215         for i in range(len(batch)):
--> 216             batch[i] -= self.global_mean.value
    217         return batch
    218 

TypeError: 'ImagePreloader' object does not support item assignment
```

it looks like model.predict() fails to apply zero-centering on the image test set

Any ideas?

**Comment by Legor** [comment]
You did not show the full code that is relevant here . From the error message one can see that you have used an ImagePreloader for your data. If you post the rest of the code it could be easier to help you.

**Comment by Staano** [comment]
Apologies Legor, 

Below is the full code:

```
training_x, training_y = image_preloader(TRAIN_DATA_FILE, image_shape=(IMG_SIZE,IMG_SIZE), 
                                        categorical_labels=True,normalize=True)
test_x, test_y = image_preloader(TEST_DATA_FILE, image_shape=(IMG_SIZE,IMG_SIZE), 
                                        categorical_labels=True,normalize=True)
valid_x, valid_y = image_preloader(VALIDATION_DATA_FILE, image_shape=(IMG_SIZE,IMG_SIZE), 
                                        categorical_labels=True,normalize=True)

img_aug = ImageAugmentation()
img_aug.add_random_flip_leftright()         
img_aug.add_random_rotation(max_angle=25.)
img_aug.add_random_blur(sigma_max=3.)

img_prep = ImagePreprocessing()
img_prep.add_featurewise_zero_center() 
img_prep.add_featurewise_stdnorm() 

# Input is a IMG_SIZE X IMG_SIZE image with 3 color channels (red, green and blue)
convnet = input_data(shape=[None, IMG_SIZE, IMG_SIZE, CHANNELS], data_preprocessing=img_prep,
                     data_augmentation=img_aug, name='input')

# Convolution#1
convnet = conv_2d(convnet, 32, 5, activation='relu')
# Max pooling
convnet = max_pool_2d(convnet, 5)

# Convolution#2
convnet = conv_2d(convnet, 64, 5, activation='relu')
# Max pooling
convnet = max_pool_2d(convnet, 5)

# Convolution#3
convnet = conv_2d(convnet, 128, 5, activation='relu')
# Max pooling
convnet = max_pool_2d(convnet, 5)

# Convolution#4
convnet = conv_2d(convnet, 128, 5, activation='relu')
# Max pooling
convnet = max_pool_2d(convnet, 5)

# Convolution#5
convnet = conv_2d(convnet, 256, 5, activation='relu')
# Max pooling
convnet = max_pool_2d(convnet, 5)

# Fully-connected 1024 node neural network
convnet = fully_connected(convnet, 1024, activation='relu')

# Dropout - throw away some data randomly during training to prevent over-fitting
convnet = dropout(convnet, 0.7)

# Fully-connected neural network with two outputs to make the final prediction
convnet = fully_connected(convnet, 2, activation='softmax')

# Tell tflearn how we want to train the network
convnet = regression(convnet, optimizer='adam', learning_rate=LEARNING_RATE, 
                     loss='categorical_crossentropy', name='targets')


model = tflearn.DNN(convnet, tensorboard_dir='/tmp/log')


MODEL_NAME = '5layers_basic_25epochs.model'
model.fit({'input': training_x}, {'targets': training_y}, n_epoch=EPOCHS, 
          validation_set=({'input': valid_x}, {'targets': valid_y}), 
          snapshot_step=500, show_metric=True, run_id=MODEL_NAME)


# Testing the Model
predY = model.predict(test_x)
```

**Comment by Legor** [comment]
Hi Staano,

First the short answer:
Change your last line (the prediction step):
`predY = model.predict(test_x)`
to
`predY = model.predict(test_x[:])`

TL;DR;
When you call 
`test_x, test_y = image_preloader(TEST_DATA_FILE, image_shape=(IMG_SIZE,IMG_SIZE), 
                                        categorical_labels=True,normalize=True)`

your test_x and test_y do not contain your data yet. 
They are but Instances of the class ImagePreloader and LabelPreloader. Only when you want to receive data from those classes (e.g. by doing test_x[1]) they will load the date (e.g images from a given path) on the fly.

The whole purpose of the Preloader classes is to be able to handle large datasets, since one may be limited by memory (RAM) and it may not be possible to load the whole dataset at once. Thats when you want to use those classes and only load a part of the data on the fly. For example during the training phase only the batches will be loaded into the RAM when using a Preloader class.

However the prediction step does not seem to support this possibility yet.

**Comment by Staano** [comment]
Thanks a lot @Legor for the explanation and support

**Comment by willduan** [comment]
@Legor great!
